File: neo/backend/src/app/api/v1/agent_mapping.py
"""API routes for agent type-to-instance mappings scoped to a workspace."""

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.dependencies import get_current_user, get_db
from app.models.user import User
from app.repositories import workspace_repository
from app.schemas.knlg_agent_mapping import (
    AgentMappingCreate,
    AgentMappingListResponse,
    AgentMappingResponse,
    AgentMappingUpdate,
)
from app.schemas.response import ApiResponse
from app.services.knlg_agent_mapping_service import KnlgAgentMappingService

router = APIRouter(
    prefix="/workspaces/{workspace_code}/agent-mappings",
    tags=["agent-mapping"],
)


def _get_workspace_id(workspace_code: str, db: Session) -> int:
    """Resolve a workspace_code to its primary key, or raise 404."""
    workspace = workspace_repository.get_workspace_by_code(db, workspace_code)
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )
    # TODO: Check workspace membership
    return int(workspace.id)


def _get_service(db: Session = Depends(get_db)) -> KnlgAgentMappingService:
    return KnlgAgentMappingService(db)


# ============ List ============


@router.get("", response_model=ApiResponse[AgentMappingListResponse])
def list_agent_mappings(
    workspace_code: str,
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    db: Session = Depends(get_db),
    service: KnlgAgentMappingService = Depends(_get_service),
    _current_user: User = Depends(get_current_user),
) -> ApiResponse[AgentMappingListResponse]:
    """List all (type -> agent_id) mappings for a workspace."""
    workspace_id = _get_workspace_id(workspace_code, db)

    items, total = service.list_mappings(
        workspace_id=workspace_id,
        page=page,
        page_size=page_size,
    )
    total_pages = (total + page_size - 1) // page_size if total > 0 else 0

    return ApiResponse.success(
        AgentMappingListResponse(
            items=[AgentMappingResponse.model_validate(m) for m in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        ),
    )


# ============ Create ============


@router.post(
    "",
    response_model=ApiResponse[AgentMappingResponse],
    status_code=status.HTTP_201_CREATED,
)
def create_agent_mapping(
    workspace_code: str,
    data: AgentMappingCreate,
    db: Session = Depends(get_db),
    service: KnlgAgentMappingService = Depends(_get_service),
    _current_user: User = Depends(get_current_user),
) -> ApiResponse[AgentMappingResponse]:
    """Create a new (type -> agent_id) mapping in the workspace."""
    workspace_id = _get_workspace_id(workspace_code, db)

    mapping = service.create_mapping(
        workspace_id=workspace_id,
        type=data.type,
        agent_id=data.agent_id,
    )
    return ApiResponse.success(AgentMappingResponse.model_validate(mapping))


# ============ Get by type ============


@router.get(
    "/{mapping_type}",
    response_model=ApiResponse[AgentMappingResponse],
)
def get_agent_mapping(
    workspace_code: str,
    mapping_type: str,
    db: Session = Depends(get_db),
    service: KnlgAgentMappingService = Depends(_get_service),
    _current_user: User = Depends(get_current_user),
) -> ApiResponse[AgentMappingResponse]:
    """Get a single mapping by its `type` within the workspace."""
    workspace_id = _get_workspace_id(workspace_code, db)
    mapping = service.get_mapping(workspace_id, mapping_type)
    return ApiResponse.success(AgentMappingResponse.model_validate(mapping))


# ============ Update ============


@router.put(
    "/{mapping_type}",
    response_model=ApiResponse[AgentMappingResponse],
)
def update_agent_mapping(
    workspace_code: str,
    mapping_type: str,
    data: AgentMappingUpdate,
    db: Session = Depends(get_db),
    service: KnlgAgentMappingService = Depends(_get_service),
    _current_user: User = Depends(get_current_user),
) -> ApiResponse[AgentMappingResponse]:
    """Update the `agent_id` of an existing mapping (type is immutable)."""
    workspace_id = _get_workspace_id(workspace_code, db)

    updated = service.update_mapping_agent(
        workspace_id=workspace_id,
        type=mapping_type,
        new_agent_id=data.agent_id,
    )
    return ApiResponse.success(AgentMappingResponse.model_validate(updated))


# ============ Delete ============


@router.delete(
    "/{mapping_type}",
    status_code=status.HTTP_200_OK,
)
def delete_agent_mapping(
    workspace_code: str,
    mapping_type: str,
    db: Session = Depends(get_db),
    service: KnlgAgentMappingService = Depends(_get_service),
    _current_user: User = Depends(get_current_user),
) -> ApiResponse[None]:
    """Delete a mapping by its `type`. Returns 404 if not found."""
    workspace_id = _get_workspace_id(workspace_code, db)
    service.delete_mapping(workspace_id, mapping_type)
    return ApiResponse.success(None, message="deleted")

File: neo/frontend/lib/api/agent-mapping.ts
/**
 * Agent Mapping API client
 *
 * Connects to backend API:
 *   GET    /api/v1/workspaces/{ws}/agent-mappings
 *   GET    /api/v1/workspaces/{ws}/agent-mappings/{type}
 *   POST   /api/v1/workspaces/{ws}/agent-mappings
 *   PUT    /api/v1/workspaces/{ws}/agent-mappings/{type}
 *   DELETE /api/v1/workspaces/{ws}/agent-mappings/{type}
 */

import { getAuthHeaders } from "@/lib/utils/auth";

const API_BASE = "";

// ============ Types ============

export interface AgentMapping {
  id: number;
  workspace_id: number;
  type: string;
  agent_id: number;
  created_at: string;
  updated_at: string;
}

export interface AgentMappingListResponse {
  items: AgentMapping[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CreateAgentMappingRequest {
  type: string;
  agent_id: number;
}

export interface UpdateAgentMappingRequest {
  agent_id: number;
}

// ============ Error handling ============

export class ApiError extends Error {
  constructor(
    message: string,
    public code: number,
    public statusCode: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<T> {
  const url = `${API_BASE}${endpoint}`;
  const authHeaders = getAuthHeaders();

  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...authHeaders,
      ...options.headers,
    },
    credentials: "include",
  });

  const data = await response.json();

  if (!response.ok) {
    throw new ApiError(
      data.message || "Request failed",
      data.code || response.status,
      response.status,
    );
  }

  if (data.code !== 0) {
    throw new ApiError(data.message, data.code, 200);
  }

  return data.data as T;
}

// ============ API functions ============

/** List all mappings in a workspace. */
export async function listAgentMappings(
  workspaceCode: string,
  params?: { page?: number; page_size?: number },
): Promise<AgentMappingListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.page_size)
    searchParams.set("page_size", params.page_size.toString());

  const qs = searchParams.toString();
  return apiRequest<AgentMappingListResponse>(
    `/api/v1/workspaces/${workspaceCode}/agent-mappings${qs ? `?${qs}` : ""}`,
  );
}

/** Get a single mapping by its `type` within the workspace. */
export async function getAgentMapping(
  workspaceCode: string,
  type: string,
): Promise<AgentMapping> {
  return apiRequest<AgentMapping>(
    `/api/v1/workspaces/${workspaceCode}/agent-mappings/${type}`,
  );
}

/** Create a new (type -> agent_id) mapping. */
export async function createAgentMapping(
  workspaceCode: string,
  data: CreateAgentMappingRequest,
): Promise<AgentMapping> {
  return apiRequest<AgentMapping>(
    `/api/v1/workspaces/${workspaceCode}/agent-mappings`,
    {
      method: "POST",
      body: JSON.stringify(data),
    },
  );
}

/** Update the agent_id of an existing mapping (type is immutable). */
export async function updateAgentMapping(
  workspaceCode: string,
  type: string,
  data: UpdateAgentMappingRequest,
): Promise<AgentMapping> {
  return apiRequest<AgentMapping>(
    `/api/v1/workspaces/${workspaceCode}/agent-mappings/${type}`,
    {
      method: "PUT",
      body: JSON.stringify(data),
    },
  );
}

/** Delete a mapping by its `type`. */
export async function deleteAgentMapping(
  workspaceCode: string,
  type: string,
): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(
    `/api/v1/workspaces/${workspaceCode}/agent-mappings/${type}`,
    {
      method: "DELETE",
    },
  );
}
